File: src/007-10001st-prime/ruby/src/euler/prime.rb
module Euler
  class Prime
    include Enumerable

    def initialize
      @primes = [2,3]
    end

    def nth(p)
      while @primes.length <= p
        self.next
      end

      @primes[p]
    end

    def next
      cur = @primes.last + 2
      while !check_prime?(cur)
        cur += 2
      end
      @primes.push(cur)
      cur
    end

    def each
      @primes.each do |p|
        yield p
      end

      while true
        yield self.next
      end
    end

    def check_prime?(n)
      max = Math.sqrt(n).floor

      @primes.each do |p|
        return true if p > max
        return false if (n % p) == 0
      end

      true
    end

    def check?(n)
      max = Math.sqrt(n).to_i

      while @primes.last <= max
        self.next
      end

      check_prime? n
    end


    def index(n)
      while n > @primes.last
        self.next
      end

      i = @primes.bsearch_index {|p, _| p >= n}
      if i.nil? || @primes[i] == n
        i
      else
        nil
      end
    end
  end
end

File: src/python/003/prime.py
import math

class Prime(object):
   def __init__(self):
      self.primes = [2, 3]
      self.prime  = 3

   def check(self, prime):
      root = math.sqrt(prime)
      for d in self.primes:
         if d > root:
            return True
         if (prime % d) == 0:
            return False
      return True

   def next(self):
      # find next prime
      self.prime += 2
      while not self.check(self.prime):
         self.prime += 2

      self.primes.append(self.prime)
      return self.primes[-1]

   def nth(self, n):
      while len(self.primes) <= n:
         self.next()
      return self.primes[n]

   def iter(self):
      for p in self.primes:
         yield p
      while True:
         yield self.next()

# generator function for primes
def iter():
   return Prime().iter()

